buku

```diff
@@ -1492,132 +1545,131 @@ class BukuDb:
             q0 += ')'
 
         try:
-            return self._fetch(q0)
+            return self._fetch(q0 + f' ORDER BY {self._order(order)}')
         except sqlite3.OperationalError as e:
             LOGERR(e)
             return []
 
+    def _search_tokens(self, keyword: str, deep=False, regex=False, markers=False):
+        """Converts a keyword into a list of tokens, based on search parameters.
+        A token is a varied-length tuple of following values: (SQL field, deep, *SQL params)."""
+        deep = not regex and deep
+        if not markers or (re.sub(r'^\*', '', keyword) and not re.match(r'^[.:>#]', keyword)):
+            s = (keyword if not markers else re.sub(r'^\*', '', keyword))
+            if not s:
+                return []
+            tags = ([s] if regex and not markers else taglist(s.split(DELIM)))
+            return [('metadata', deep, s), ('url', deep, s), ('desc', deep, s)] + (tags and [('tags', deep, *tags)])
+        if re.match(r'^\..', keyword):  # checking prefix + ensuring keyword[1:] is not empty
+            return [('metadata', deep, keyword[1:])]
+        if re.match(r'^:.', keyword):
+            return [('url', deep, keyword[1:])]
+        if re.match(r'^>.', keyword):
+            return [('desc', deep, keyword[1:])]
+        if re.match(r'^#,?[^,]', keyword):
+            tags = ([re.sub(r'^#,?', '', keyword)] if regex else taglist(keyword[1:].split(DELIM)))
+            return tags and [('tags', not keyword.startswith('#,'), *tags)]
+        return []
+
+    def _search_clause(self, tokens, regex=False) -> Tuple[str, List[str]]:
+        """Converts a list of tokens into an SQL clause. (See also: BukuDb._search_tokens().)
+        If regex is True, the token is treated as a raw regex and the paired deep parameter is ignored."""
+        border = lambda k, c: (',' if k == 'tags' else r'\b' if c.isalnum() else '')
+
+        args, clauses = [], []
+        if regex:
+            for field, deep, param in tokens:
+                clauses += [field + ' REGEXP ?']
+                args += [param]
+        else:
+            for field, deep, *params in tokens:
+                _clauses = []
+                for param in params:
+                    if deep:
+                        _clauses += [field + " LIKE ('%' || ? || '%')"]
+                    else:
+                        _clauses += [field + ' REGEXP ?']
+                        param = border(field, param[0]) + re.escape(param) + border(field, param[-1])
+                    args += [param]
+                clauses += (_clauses if len(_clauses) < 2 else [f'({" AND ".join(_clauses)})'])
+        return ' OR '.join(clauses), args
```

Comment:
This is based on the original implementation of `searchdb()`; `regex=True` means that `deep` is ignored and only one `param` is expected in a token (since comma-splitting is not done in regex mode), and otherwise we rely on `\b` to determine the word edge (or `,` when working with tags).

Note that I removed stripping `/` from the end of the string, since it's easy to not include a slash for the user if it's not necessary, but if it's stripped then it's impossible to actually search for it correctly (e.g. specifying `.com/` in URL search may not produce the desired results).